content/blog/httr2-1-0-0/index.Rmd

```diff
@@ -0,0 +1,268 @@
+    `req_cache()` automatically prunes the cache, ensuring that by default it stays under 1 GB
+
+-   `req_throttle()` automatically adds a small delay before each request so you can avoid hammering a server with many requests.
+
+-   `req_progress()` adds a progress bar for long downloads or uploads.
+
+-   `req_cookie_preserve()` lets you preserve cookies across requests.
+
+Additionally, httr2 provides many helpers for authenticating with OAuth, wrapping many more styles than httr.
+You've probably used OAuth a bunch without knowing what it's called: you use it when you login to a non-Google website using your Google account, when you give your phone access to your twitter account, or when you login to a streaming app on your smart TV.
+OAuth is a big, complex topic, and is documented in `vignette("oauth2")`
+
+## Multiple requests
+
+httr2 includes three functions to perform multiple requests:
+
+-   `req_perform_sequential()` takes a list of requests and performs them in order.
+
+-   `req_perform_parallel()` takes a list of requests and performs them in parallel (up to 6 at a time by default).
+    It's similar to `req_perform_sequential()`, but is obviously faster, at the expense of potentially hammering a server.
+    However, it also has some limitations: most importantly it can't re-request an expired OAuth token and it doesn't respect `req_retry()` or `req_throttle()`.
+
+-   `req_perform_iterative()` takes a single request and a callback function to generate the next request from previous response.
+    It'll keep going until it either your callback function returns `NULL` or the `max_reqs` requests have been performed.
+
+For example, imagine we wanted to download each person from the [Star Wars API](https://swapi.dev). The urls have a very consistent structure so we can generate a bunch of them, then create the corresponding requests.
+
+```{r}
+urls <- paste0("https://swapi.dev/api/people/", 1:10)
+reqs <- lapply(urls, request)
+```
+
+Now I can perform those requests, collecting a list of responses:
+
+```{r}
+resps <- req_perform_sequential(reqs)
+```
+
+These responses contain their data in a JSON body:
+
+```{r}
+resps |> 
+  _[[1]] |> 
+  resp_body_json() |> 
+  str()
+```
+
+There's lots of ways to deal with this sort of data (e.g. for loops or functional programming), httr2 comes with a helper,`resps_data()`.
+It takes a callback function that retrieves the data for each response then concatenate all the responses back together.
+In this case that means we need to wrap `resp_body_json()` in a list, so we get one list for each person, rather than one list in total:
+
+```{r}
+resps |> 
+  resps_data(\(resp) list(resp_body_json(resp))) |> 
+  _[1:3] |> 
+  str(list.len = 10)
+```
+
+Another option would be to convert each response into a data frame or tibble.
+That's a little tricky here because of the lists that will need to become list-columns[^2], so we'll avoid that challenge here by focussing the first nine columns:
```

Comment:
```suggestion
That's a little tricky here because of the lists that will need to become list-columns[^2], so we'll avoid that challenge here by focussing on the first nine columns:
```